### PR TITLE
[serve] Add Serve CLI commands for REST API

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This changes adds Serve CLI commands that wrap the REST API. With these commands, users can write their deployment configurations in standalone YAML files. The CLI translates these files to JSON and applies them to a Ray cluster via Serve's REST API.

## Related issue number

These changes are related to #21881.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - These changes add new unit tests to `test_cli.py`.

### CODE REVIEW

File: python/ray/serve/scripts.py [hunk 1]
```diff
@@ -114,9 +129,68 @@ class may or may not be decorated with ``@serve.deployment``.
...
+    help='Address of the Ray dashboard to query. For example, "http://localhost:8265".',
+)
+def deploy(config_file_name: str, address: str):
+    full_address_path = f"{address}/api/serve/deployments/"
+
+    with open(config_file_name, "r") as config_file:
+        config = yaml.safe_load(config_file)
+
+    # Generate a schema using the config to ensure its format is valid
+    ServeInstanceSchema.parse_obj(config)
```
Comment (edoakes) [inline python/ray/serve/scripts.py:163]: how bad is the output we get here if invalid?

File: python/ray/serve/scripts.py [hunk 2]
```diff
@@ -114,9 +129,68 @@ class may or may not be decorated with ``@serve.deployment``.
...
+
+    response = requests.put(full_address_path, json=config)
+
+    if response.status_code == 200:
+        print(
```
Comment (edoakes) [inline python/ray/serve/scripts.py:168]: use logger not print

Comment (edoakes) [reply python/ray/serve/scripts.py:168]: see the shared Ray CLI logger we use in the jobs cli `dashboard/modules/job/cli.py`

File: python/ray/serve/scripts.py [hunk 3]
```diff
@@ -114,9 +129,68 @@ class may or may not be decorated with ``@serve.deployment``.
...
+    default=os.environ.get("RAY_ADDRESS", "http://localhost:8265"),
+    required=False,
+    type=str,
+    help='Address of the Ray dashboard to query. For example, "http://localhost:8265".',
+)
+def info(address: str):
+    full_address_path = f"{address}/api/serve/deployments/"
+    response = requests.get(full_address_path)
+    if response.status_code == 200:
+        print(response.json())
```
Comment (edoakes) [inline python/ray/serve/scripts.py:194]: do we need to prettyprint this?

**Review by edoakes — CHANGES_REQUESTED** [review]

File: python/ray/serve/scripts.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/scripts.py]: looks ok for now

File: python/ray/serve/scripts.py (hunk 3 above)
Comment (shrekris-anyscale) [reply python/ray/serve/scripts.py]: Good idea, I stylized the output.

File: python/ray/serve/scripts.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/scripts.py]: Fixed.

File: dashboard/modules/serve/serve_head.py [hunk 4]
```diff
@@ -58,9 +53,9 @@ async def delete_serve_application(self, req: Request) -> Response:
     @routes.put("/api/serve/deployments/")
     @optional_utils.init_ray_and_catch_exceptions(connect_to_serve=True)
     async def put_all_deployments(self, req: Request) -> Response:
-        serve_application_json = await req.json()
+        serve_application_text = await req.text()
         serve_application_schema = ServeApplicationSchema.parse_raw(
```
Comment (edoakes) [inline dashboard/modules/serve/serve_head.py:57]: huh, why doesn't this take json?

File: python/ray/serve/scripts.py [hunk 5]
```diff
@@ -14,6 +15,20 @@
...
+from ray.autoscaler._private.cli_logger import cli_logger
+
+
+def log_failed_request(response: requests.models.Response):
+    error_message = (
+        "\nRequest failed. Got response status code "
+        f"{response.status_code} with the following message:"
+        f"\n\n{response.text}"
+    )
+    print(error_message)
```
Comment (edoakes) [inline python/ray/serve/scripts.py:28]: why print and `cli_logger.error`?

**Review by edoakes — APPROVED** [review]
Looks good, I think you just left an errant `print` in.

File: python/ray/serve/scripts.py (hunk 5 above)
Comment (shrekris-anyscale) [reply python/ray/serve/scripts.py]: Good catch, I missed that when I was replacing print statements with the logger. Fixed.

File: dashboard/modules/serve/serve_head.py (hunk 4 above)
Comment (shrekris-anyscale) [reply dashboard/modules/serve/serve_head.py:57]: It does take json, but I left it in text form there. `ServeApplicationSchema.parse_raw()` should convert the text into the schema.

**Comment by edoakes** [comment]
@shrekris-anyscale test failed

**Pull request merged**